### PR TITLE
added JSON as external package

### LINKED ISSUE #345: JSON should be on STIR dependencies
This is to allow conversion from HU to mu-values

### PR DESCRIPTION
**Pull request by paskino**
closes #345

### CODE REVIEW

File: SuperBuild/External_JSON.cmake [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ExternalProject_Include_Dependencies(${proj} DEPENDS_VAR ${proj}_DEPENDENCIES)
+
+# Set external name (same as internal for now)
+set(externalProjName ${proj})
+
+set(${proj}_SOURCE_DIR "${SOURCE_ROOT_DIR}/${proj}" )
+set(${proj}_BINARY_DIR "${SUPERBUILD_WORK_DIR}/builds/${proj}/build" )
+set(${proj}_DOWNLOAD_DIR "${SUPERBUILD_WORK_DIR}/downloads/${proj}" )
+set(${proj}_STAMP_DIR "${SUPERBUILD_WORK_DIR}/builds/${proj}/stamp" )
+set(${proj}_TMP_DIR "${SUPERBUILD_WORK_DIR}/builds/${proj}/tmp" )
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: please replace all this by the appropriate call to a macro. I've done some clean-up work...

File: SuperBuild/External_JSON.cmake [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  ExternalProject_Add(${proj}
+    ${${proj}_EP_ARGS}
+    ${${proj}_EP_ARGS_GIT}
+    SOURCE_DIR ${${proj}_SOURCE_DIR}
+    BINARY_DIR ${${proj}_BINARY_DIR}
+    DOWNLOAD_DIR ${${proj}_DOWNLOAD_DIR}
+    STAMP_DIR ${${proj}_STAMP_DIR}
+    TMP_DIR ${${proj}_TMP_DIR}
+    INSTALL_DIR ${libcilreg_Install_Dir}
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: all this needs to be replaced as well. was using `cil` anyway

File: SuperBuild/External_JSON.cmake [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    # apparently this is the only way to pass environment variables to
+    # external projects
+    CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env CIL_VERSION=${${proj}_TAG}
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: surely not

File: SuperBuild/External_JSON.cmake [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        -DCMAKE_BUILD_TYPE:STRING=Release
+        
+    INSTALL_COMMAND ${CMAKE_COMMAND} --build . --target install
+    DEPENDS ${${proj}_DEPENDENCIES}
+  )
+
+
+  set(${proj}_ROOT        ${${proj}_SOURCE_DIR})
+  set(${proj}_INCLUDE_DIR ${${proj}_SOURCE_DIR})
+
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: no point in doing this

File: SuperBuild/External_JSON.cmake (hunk 1 above)
Comment (paskino) [reply SuperBuild/External_JSON.cmake]: done in the latest commit

File: SuperBuild/External_JSON.cmake [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+SetCanonicalDirectoryNames(${proj})
+# Get any flag from the superbuild call that may be particular to this projects CMAKE_ARGS
+SetExternalProjectFlags(${proj})
+
+if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalProjName}}" ) )
+  message(STATUS "${__indent}Adding project ${proj}")
+
+  ### --- Project specific additions here
+  set(${proj}_Install_Dir ${SUPERBUILD_INSTALL_DIR})
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: not used. (do check)

File: SuperBuild/External_JSON.cmake [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  set(CMAKE_LIBRARY_PATH ${CMAKE_LIBRARY_PATH} ${SUPERBUILD_INSTALL_DIR})
+  set(CMAKE_INCLUDE_PATH ${CMAKE_INCLUDE_PATH} ${SUPERBUILD_INSTALL_DIR})
+
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: not used

File: SuperBuild/External_JSON.cmake [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  message("${proj} URL " ${${proj}_URL}  ) 
+  message("${proj} TAG " ${${proj}_TAG}  ) 
+
+  # Sets ${proj}_URL_MODIFIED and ${proj}_TAG_MODIFIED
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: not anymore... probably just remove comment

File: SuperBuild/External_JSON.cmake [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    ${${proj}_EP_ARGS}
+    ${${proj}_EP_ARGS_GIT}
+    ${${proj}_EP_ARGS_DIRS}
+    DEPENDS ${${proj}_DEPENDENCIES}
+  )
+
+
+  else()
+    ExternalProject_Add_Empty(${proj} DEPENDS "${${proj}_DEPENDENCIES}"
+      SOURCE_DIR ${${proj}_SOURCE_DIR}
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]: `${${proj}_EP_ARGS_DIRS}`

**Comment by paskino** [comment]
Hopefully this addresses all the points raised.

File: SuperBuild/External_JSON.cmake (hunk 6 above)
Comment (KrisThielemans) [reply SuperBuild/External_JSON.cmake]: why resolved? Are these variables used?

File: SuperBuild/External_JSON.cmake (hunk 7 above)
Comment (KrisThielemans) [reply SuperBuild/External_JSON.cmake]: why resolved? The (wrong) comment line is still there

File: SuperBuild/External_JSON.cmake [hunk 9]
```diff
@@ -0,0 +1,58 @@
+#========================================================================
+# Author: Edoardo Pasca
+# Copyright 2018-2019 STFC
```
Comment (KrisThielemans) [inline SuperBuild/External_JSON.cmake]:
```suggestion
# Copyright 2018-2020 STFC
# Copyright 2018-2020 University College London
```

**Comment by paskino** [comment]
nice and clean now?

**Comment by KrisThielemans** [comment]
👍

**Comment by paskino** [comment]
It could be used as template for CMake external projects.

**Comment by KrisThielemans** [comment]
A lot of them are close to this already, but not CCPi ones. sorry :-)

In fact, there's still some repetition after that, so with a little bit of thought this can be made into another macro. The current example is (too) simple, so we just need to think a bit on how to allow some flexibility. We'd need to allow for extra `CMAKE_ARGS` for instance.

Left as an exercise to the reader...

**Pull request merged**